File: app/src/main/java/com/sortedqueue/programmercreek/adapter/CodeEditorRecyclerAdapter.java
package com.sortedqueue.programmercreek.adapter;

import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.Html;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.sortedqueue.programmercreek.R;
import com.sortedqueue.programmercreek.util.PrettifyHighlighter;

import java.util.ArrayList;

import butterknife.Bind;
import butterknife.ButterKnife;

/**
 * Created by Alok on 12/04/17.
 */
public class CodeEditorRecyclerAdapter extends RecyclerView.Adapter<CodeEditorRecyclerAdapter.ViewHolder> {

    private Context context;
    private ArrayList<String> programLines;
    private PrettifyHighlighter prettifyHighlighter;
    private String programLanguage;
    private int editableIndex = -1;
    private String TAG = CodeEditorRecyclerAdapter.class.getSimpleName();

    public CodeEditorRecyclerAdapter(Context context, ArrayList<String> programLines, String programLanguage) {
        this.context = context;
        this.programLines = programLines;
        this.programLanguage = programLanguage;
        this.prettifyHighlighter = PrettifyHighlighter.getInstance();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_edit_code, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        String programLine = programLines.get(position);
        //holder.codeEditText.setEnabled(position == editableIndex);
        if( programLine.contains("<") || programLine.contains(">")) {
            holder.codeEditText.setText(programLine);
            holder.codeEditText.setTextColor(Color.parseColor("#006699"));
        }
        else {
            if(Build.VERSION.SDK_INT >= 24) {
                holder.codeEditText.setText(Html.fromHtml(prettifyHighlighter.highlight(programLanguage, programLine), Html.FROM_HTML_MODE_LEGACY));
            }
            else {
                holder.codeEditText.setText(Html.fromHtml(prettifyHighlighter.highlight(programLanguage, programLine)));
            }
        }
    }

    @Override
    public int getItemCount() {
        return programLines.size();
    }

    public String getCode() {
        String code = "";
        for( String programLine : programLines ) {
            code += programLine + "\n";
        }
        return code;
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, TextWatcher {

        @Bind(R.id.codeEditText)
        EditText codeEditText;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            codeEditText.setOnClickListener(this);
            codeEditText.addTextChangedListener(this);
        }

        @Override
        public void onClick(View v) {
            editableIndex = getAdapterPosition();
        }

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable s) {
            if( editableIndex == -1 ) {
                return;
            }
            String programLine = s.toString().trim();
            Log.d(TAG, "Setting line : " + programLine);
            if( programLine.trim().length() > 0 ) {
                int position = getAdapterPosition();
                programLines.set(position, programLine);

                if( position != RecyclerView.NO_POSITION ) {
                    if( programLine.contains("<") || programLine.contains(">")) {
                        codeEditText.setText(programLine);
                        codeEditText.setTextColor(Color.parseColor("#006699"));
                    }
                    if(Build.VERSION.SDK_INT >= 24) {
                        codeEditText.setText(Html.fromHtml(prettifyHighlighter.highlight(programLanguage, programLine), Html.FROM_HTML_MODE_LEGACY));
                    }
                    else {
                        codeEditText.setText(Html.fromHtml(prettifyHighlighter.highlight(programLanguage, programLine)));
                    }
                }

            }

        }
    }

}

File: app/src/main/java/com/sortedqueue/programmercreek/adapter/CodeEditorRecyclerAdapter.kt
package com.sortedqueue.programmercreek.adapter

import android.content.Context
import android.graphics.Color
import android.os.Build
import android.support.v7.widget.RecyclerView
import android.text.Editable
import android.text.Html
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText

import com.sortedqueue.programmercreek.R
import com.sortedqueue.programmercreek.util.PrettifyHighlighter

import java.util.ArrayList




/**
 * Created by Alok on 12/04/17.
 */
class CodeEditorRecyclerAdapter : RecyclerView.Adapter<CodeEditorRecyclerAdapter.ViewHolder> {

    private var context: Context? = null
    var programLines: ArrayList<String>? = null
        private set
    private var prettifyHighlighter: PrettifyHighlighter? = null
    private var programLanguage: String? = null
    private var editableIndex = -1
    private val TAG = CodeEditorRecyclerAdapter::class.java.simpleName
    private var codeMode = false

    constructor(context: Context, programLines: ArrayList<String>, programLanguage: String) {
        this.context = context
        this.programLines = programLines
        this.programLanguage = programLanguage
        this.prettifyHighlighter = PrettifyHighlighter.instance
    }

    constructor(context: Context, programLines: ArrayList<String>, programLanguage: String, codeMode: Boolean) {
        this.context = context
        this.codeMode = codeMode
        this.programLines = programLines
        this.programLanguage = programLanguage
        this.prettifyHighlighter = PrettifyHighlighter.instance
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(
                if (codeMode)
                    R.layout.item_code
                else
                    R.layout.item_edit_code, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val programLine = programLines!![position]
        //holder.codeEditText.setEnabled(position == editableIndex);
        if (programLine.contains("<") || programLine.contains(">")) {
            holder.codeEditText!!.setText(programLine)
            holder.codeEditText!!.setTextColor(Color.parseColor("#006699"))
        } else {
            if (Build.VERSION.SDK_INT >= 24) {
                holder.codeEditText!!.setText(Html.fromHtml(prettifyHighlighter!!.highlight(programLanguage!!, programLine), Html.FROM_HTML_MODE_LEGACY))
            } else {
                holder.codeEditText!!.setText(Html.fromHtml(prettifyHighlighter!!.highlight(programLanguage!!, programLine)))
            }
        }
    }

    override fun getItemCount(): Int {
        return programLines!!.size
    }

    val code: String
        get() {
            var code = ""
            for (programLine in programLines!!) {
                code += programLine + "\n"
            }
            return code
        }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener, TextWatcher {

        internal var codeEditText: EditText = itemView.findViewById(R.id.codeEditText) as EditText

        init {

            codeEditText!!.isEnabled = false
            codeEditText!!.setOnClickListener(this)
            codeEditText!!.addTextChangedListener(this)
        }

        override fun onClick(v: View) {

        }


        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

        }

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            editableIndex = adapterPosition
        }

        override fun afterTextChanged(s: Editable) {
            if (editableIndex == -1) {
                return
            }
            val position = adapterPosition
            if (position != RecyclerView.NO_POSITION) {
                val programLine = s.toString().trim { it <= ' ' }
                Log.d(TAG, "Setting line : " + programLine)
                if (programLine.trim { it <= ' ' }.isNotEmpty()) {
                    programLines!![position] = programLine
                } else {
                    programLines!![position] = ""
                }
            }

        }
    }

}
